ppcs/ppc0022-metaprogramming.md

```diff
@@ -0,0 +1,385 @@
+
+**Optionally on globs** If globs are also supported, then the `$name` can be expressed without a sigil (or maybe the `*` sigil?) to  create and return a metaglob instead of a metasymbol.
+
+#### `remove_symbol`
+
+```perl
+$metapackage->remove_symbol($name);
+```
+
+Removes a symbol. Nothing is returned.
+
+#### `list_symbols`
+
+```perl
+@metasymbols = $metapackage->list_symbols($filter)
+```
+
+Returns a list of symbols in the given package, in no particular order. (And in particular, we make no guarantees about the order of the results as compared to the order of the source code which declared it).
+
+TODO: Define what the filters look like. Need to be able to select scalar/array/hash/code as well as sub-stashes. Actually, do we need sub-stashes if we have `list_packages`?
+
+**Optionally on globs** If globs are also supported, this raises the fun question of whether the returned list contains the actual globs or only the symbols from within them. Perhaps there should be another method `->list_globs` that yields just globs?
+
+### Methods on Meta-Symbol Objects
+
+Several of the above methods return instances of a meta-symbol object. In practice, any meta-symbol object will represent either a variable (when using the "$@%" sigils), or a subroutine (when using the "&"). Each of these cases is described further below. The following methods are available on both kinds.
+
+#### `basename`
+
+```perl
+$name = $metasymbol->basename;
+```
+
+Returns the final part of the symbol's fully qualified name; the name within the package after the final `::` but **excluding** the leading sigil.
+
+#### `sigil`
+
+```perl
+$sigil = $metasymbol->sigil;
+```
+
+Returns the single-character sigil that leads the symbol's name. This will be one of "$", "@", "%" for variables or "&" for subroutines.
+
+#### `package`
+
+```perl
+$metapackage = $metasymbol->package;
+```
+
+Returns the meta-package object instance representing the package to which this symbol belongs.
+
+#### `name`
+
+```perl
+$name = $metasymbol->name;
+```
+
+Returns the fully-qualified name for this symbol, including its sigil and package name. This would be equivalent to
+
+```perl
+$name = $metasymbol->sigil . join "::", $metasymbol->package->name, $metasymbol->basename;
+```
+
+#### `is_`*\**
+
+```perl
+$bool = $metasymbol->is_scalar;
+$bool = $metasymbol->is_array;
+$bool = $metasymbol->is_hash;
+$bool = $metasymbol->is_subroutine;
+```
+
+Returns true in each of the four cases where the object represents the given type of symbol, or false in the other three cases.
+
+### Methods on Meta-Glob Objects
+
+If globs are represented directly, the following methods would be available on them.
+
+#### `basename`
+
+#### `package`
+
+#### `name`
+
+Similar behaviour to on meta-symbols.
+
+#### `can_scalar`, `get_scalar`
```

Comment:
The motivation for `can` came from `$obj->can( 'method' )` either returning a coderef or undef.

I also felt that having a "get" form that throws an exception rather than returning undef makes for cleaner-looking code if you're not expecting to find things missing.

```perl
$pkg->get_package( "subpackage" )->get_symbol( '&funcname' )->get_whatever ...
```

If any of those are missing you get more obvious errors of the form "Package Thing does not contain a 'subpackage' package" rather than "Can't call method \"get_package\" on an undefined value at ...". Otherwise, you end up writing long chains like this with lots of manual error-checking:

```perl
(($pkg->get_package('subpackage') or die "Has no subpackage")->get_symbol('&func') or die "Has no func"))->...
```

It's far from a concern specifically about this meta stuff specifically, and more a more general consideration of what feels best for any kind of API with "get me a thing that might not exist" query functions in it, basically.

Reply:
`can` is named as such because it checks for methods. Methods are actions, so you check if something can do them. It's the wrong verb for this API.

Reply:
Oh I see - yes from that angle `can` is probably the wrong word here. I'm loathe to suggest `has` though, simply because I'd imagine a `has` would just return a true/false boolean. Whereas we're looking at

```perl
$metathing_or_undef = $metapkg->has_symbol( 'NAME' );
$metathing_or_die   = $metapkg->get_symbol( 'NAME' );
```

currently. I don't have any other good verb suggestions; unless we go for something more wordy. The following would make it very clear what the behaviour is (and also solve similar questions for add/remove) but seems probably overly verbose in the common case.

```perl
$metathing = $metapkg->add_if_missing( 'NAME' ); # throws if already present
$metathing = $metapkg->add_or_return( 'NAME' ); # silently returns it if already present

$metathing = $metapkg->get_or_undef( 'NAME' ); # returns undef if missing
$metathing = $metapkg->get_or_fail( 'NAME' ); # dies if missing

$metapkg->remove_or_fail( 'NAME' ); # dies if missing
$metapkg->remove_if_present( 'NAME' ); # silently ignores if missing
```

These names mostly suck. I was inspired by SQL's `INSERT OR UPDATE ...` but that fits the more wordy SQL.

As you can tell this part is still very much a WIP here ;)